src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -21,489 +21,211 @@
-  }
-
-  // update pattern history if this is a new record
-  if (recordNumHistory_.size() == 0 || recordNum > lastRecordNum) {
-    patternNZHistory_.emplace_back(patternNZ.begin(), patternNZ.end());
-    recordNumHistory_.push_back(recordNum);
-    if (patternNZHistory_.size() > maxSteps_) {
-      patternNZHistory_.pop_front();
-      recordNumHistory_.pop_front();
-    }
-  }
-
-  // if input pattern has greater index than previously seen, update
-  // maxInputIdx and augment weight matrix with zero padding
-  if (patternNZ.size() > 0) {
-    const UInt maxInputIdx = *max_element(patternNZ.begin(), patternNZ.end());
-    if (maxInputIdx > maxInputIdx_) {
-      maxInputIdx_ = maxInputIdx;
-    }
-  }
-
-  // if in inference mode, compute likelihood and update return value
-  if (infer) {
-    infer_(patternNZ, actValueList, result);
-  }
-
-  // update weights if in learning mode
-  if (learn) {
-    for (size_t categoryI = 0; categoryI < bucketIdxList.size(); categoryI++) {
-      const UInt bucketIdx = bucketIdxList[categoryI];
-      const Real64 actValue = actValueList[categoryI];
-      // if bucket is greater, update maxBucketIdx_ and augment weight
-      // matrix with zero-padding
-      if (bucketIdx > maxBucketIdx_) {
-        maxBucketIdx_ = bucketIdx;
-      }
-
-      // update rolling averages of bucket values
-      while (actualValues_.size() <= maxBucketIdx_) {
-        actualValues_.push_back(0.0);
-        actualValuesSet_.push_back(false);
-      }
-      if (!actualValuesSet_[bucketIdx] || category) {
-        actualValues_[bucketIdx] = actValue;
-        actualValuesSet_[bucketIdx] = true;
-      } else {
-        actualValues_[bucketIdx] =
-            ((1.0 - actValueAlpha_) * actualValues_[bucketIdx]) +
-            (actValueAlpha_ * actValue);
-      }
-    }
-
-    // compute errors and update weights
-    auto patternIteration = patternNZHistory_.begin();
-    for (auto learnRecord = recordNumHistory_.begin();
-         learnRecord != recordNumHistory_.end();
-         learnRecord++, patternIteration++) {
-      const vector<UInt> learnPatternNZ = *patternIteration;
-      const UInt nSteps = recordNum - *learnRecord;
-
-      // update weights
-      if (binary_search(steps_.begin(), steps_.end(), nSteps)) {
-        const vector<Real64> error = calculateError_(bucketIdxList, learnPatternNZ, nSteps);
-        Matrix& w = weightMatrix_.at(nSteps);
-	NTA_ASSERT(alpha_ > 0.0);
-        for (const auto& bit : learnPatternNZ) {
-          for(size_t i = 0; i < error.size(); i++) {
-            const auto val = get_(w, bit, i) + alpha_ * error[i];
-	    if(val == 0) continue;
-	    w[bit][i] = val;
-          }
-        }
-      }
-    }
-  }
+PDF Classifier::infer(const SDR & pattern)
+{
+  // Check input dimensions, or if this is the first time the Classifier has
+  // been used then initialize it with the given SDR's dimensions.
+  if( dimensions_.empty() ) {
+    dimensions_ = pattern.dimensions;
```

Comment:
dimensions cannot change during runtime, let's set them during init (will be able to move this whole if-elseif into initialize)

Reply:
This way the user doesn't need to tell us the dimensions ahead of time.  It's one less argument to remember.